django/db/models/manager.py

```diff
@@ -242,12 +244,23 @@ class Manager(BaseManager.from_queryset(QuerySet)):
 class ManagerDescriptor(object):
     # This class ensures managers aren't accessible via model instances.
     # For example, Poll.objects works, but poll_obj.objects raises AttributeError.
-    def __init__(self, manager):
+    # It also ensures managers defined on non-abstract base classes are not inherited by child classes.
+    def __init__(self, manager, model):
         self.manager = manager
+        self.model = model
 
     def __get__(self, instance, cls=None):
         if instance is not None:
             raise AttributeError("Manager isn't accessible via %s instances" % cls.__name__)
+
+        if not (cls == self.model or (cls._meta.proxy and cls._meta.concrete_model is None)):
```

Comment:
A comment for why the check for `(cls._meta.proxy and cls._meta.concrete_model is None)` is needed would be helpful, since it is not very obvious. Mainly, the `cls._meta.concrete_model is None` doesn't sound correct at first. From what I understand, we need this check because the descriptor is accessed in `ModelBase` when to copy the manager for proxy models, and at that point `cls._meta.concrete_model` has not been set.